### PR TITLE
JAMES-1711 setMailboxes : sort requests

### PR DESCRIPTION
**Pull request by fvn-linagora**
Create a tree of Mailbox objects based on create / update / destroy field of setMailboxes method. Once this tree is created, sort the tree in a root-to-leaf ordering using a specialized library. Also handle reverse ordering (for update and destroy)

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: where does this license come from ? it's not the usual one

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    public void registerItem(T item) {
+        builder.addVertex(item);
+        T parentNode = getParent.apply(item);
+        if (parentNode != null) {
+            builder.addEdge(parentNode, item);
+        }
+    }
+
+    public Stream<T> getOrdered() {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: how am I supposed to now that you mean root-to-leaf order ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        DefaultDirectedGraph<T, DefaultEdge> graph = builder.build();
+        AbstractGraphIterator<T, DefaultEdge> i = new TopologicalOrderIterator<>(graph);
+        Iterable<T> iterable = () -> i;
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: could you extract these two lines to something meaningful ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+public class DependencyGraph<T> {
+
+
+    private final DirectedGraphBuilder<T, DefaultEdge, DefaultDirectedGraph<T, DefaultEdge>> builder;
+
+    private final Function<T, String> getName;
+    private final Function<T, T> getParent;
+
+
+    public DependencyGraph(Function<T, String> getName, Function<T, T> getParent) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: getParent should be `Function<T, Optional<T>>`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import org.jgrapht.graph.builder.DirectedGraphBuilder;
+import org.jgrapht.traverse.AbstractGraphIterator;
+import org.jgrapht.traverse.TopologicalOrderIterator;
+
+public class DependencyGraph<T> {
+
+
+    private final DirectedGraphBuilder<T, DefaultEdge, DefaultDirectedGraph<T, DefaultEdge>> builder;
+
+    private final Function<T, String> getName;
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: getName is never used

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Map;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.apache.james.jmap.model.mailbox.Mailbox;
+
+public class MailboxPathSorter {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: Where does `Path` come from ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    public Stream<Mailbox> sortFromRootToLeaf(List<Mailbox> mailboxes) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: List to Stream ? Why not Stream to Stream ? or List to List ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        Map<String, Mailbox> mapOfMailboxesById = mailboxes.stream()
+                .collect(Collectors.toMap(x -> x.getId(), x -> x));
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: replace x -> x by Function::identity

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    private Mailbox getParentMailbox(Mailbox m, Map<String, Mailbox> map) {
+        return m.getParentId()
+                .map(p -> map.get(p))
+                .orElse(null);
+    }
+
+    public Stream<Mailbox> sortFromLeafToRoot(List<Mailbox> mailboxes) {
+        Iterator<Mailbox> mailboxIterator = sortFromRootToLeaf(mailboxes)
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: why do you want to use an iterator when you really want a list then a stream ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        DependencyGraph<FakeCommitNode> graph = new DependencyGraph<>(FakeCommitNode::getHash, m -> null);
+        ImmutableList<FakeCommitNode> isolatedMailboxes = ImmutableList.of(new FakeCommitNode("A"), new FakeCommitNode("B"), new FakeCommitNode("C"));
+        isolatedMailboxes.stream().forEach(graph::registerItem);
+
+        List<FakeCommitNode> orderedResultList = graph.getOrdered().collect(Collectors.toList());
+
+        assertThat(orderedResultList).isEqualTo(isolatedMailboxes);
+    }
+
+    private static class FakeCommitNode {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: what is this concept ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        private final Optional<String> message;
+        private final Optional<FakeCommitNode> parent;
+
+        @VisibleForTesting
+        FakeCommitNode(String message) {
+            this(Optional.ofNullable(message), Optional.empty());
+        }
+
+        @VisibleForTesting
+        FakeCommitNode(Optional<String> message, Optional<FakeCommitNode> parent) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: why a nullable message ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+public class DependencyGraphTest {
+
+    @Test
+    public void getOrderedShouldReturnOrderedMailbox() {
+        FakeCommitNode a = new FakeCommitNode("A");
+        FakeCommitNode b = new FakeCommitNode(Optional.of("B"), Optional.of(a));
+        FakeCommitNode c = new FakeCommitNode(Optional.of("C"), Optional.of(b));
+
+        Map<String, FakeCommitNode > mapOfMailboxesById = ImmutableList.of(b, a, c).stream()
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: the following lines look like a logic duplication from MailboxPathSorter

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Test
+    public void getOrderedWithEmptyGraphShouldReturnEmpty() {
+        DependencyGraph<FakeCommitNode> graph = new DependencyGraph<>(FakeCommitNode::getHash, m -> null);
+        assertThat(graph.getOrdered()).isEmpty();
+    }
+
+    @Test
+    public void getOrderedOnIsolatedVerticesShouldReturnSameOrder() {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: no test on two isolated trees ? on bigger trees ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        Mailbox inbox = Mailbox.builder().name("INBOX").id("INBOX").build();
+        Mailbox a = Mailbox.builder().name("A").id("A").parentId("INBOX").build();
+        Mailbox b = Mailbox.builder().name("B").id("B").parentId("INBOX").build();
+        Mailbox c = Mailbox.builder().name("C").id("C").parentId("B").build();
+        Mailbox d = Mailbox.builder().name("D").id("D").parentId("A").build();
+        Mailbox e = Mailbox.builder().name("E").id("E").parentId("C").build();
+
+        MailboxPathSorter sut = new MailboxPathSorter();
+        ImmutableList<Mailbox> input = ImmutableList.of(b, c, d, a, inbox, e);
+        List<String> result = sut.sortFromRootToLeaf(input)
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java]: what is the action and what is the property extraction ? why don't you separate properties extraction from action ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+                .map(Mailbox::getName)
+                .collect(Collectors.toList());
+
+        assertThat(result).containsExactly("INBOX", "B", "A", "C", "D", "E");
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java]: what if the library suddenly return "A" before "B" ? Would it be correct ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 1 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: content is the same. Only copyright notice formatting had been lost from my IDE project settings.
Fixed.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 2 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: renamed to getBuildChain (suggesting that the order is appropriate for building from scratch)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 3 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: extracted method getStreamFromIterator

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 4 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: changed

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 2 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: I don't understand what you mean. Could you elaborate ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 5 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: Removed.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java (hunk 7 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: takes List and returns List

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java (hunk 8 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: replaced, but I still don't get it.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 2 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: Usually build chains graphs are similar to dependencies graph but with edges directions inverted. 
eg. if A depends on B, then B should be built first, then A. So the chain would be B->A (note the arrow no longer means depends on here)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java (hunk 6 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: Would MailboxHierarchySorter be OK ?
MailboxSorter sounds more generic than it actually is.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java (hunk 9 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: previously fixed when changing returned type from Stream to List

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 11 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: removed Optional

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java (hunk 14 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java]: moved property extraction to assertion part

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java (hunk 15 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/MailboxPathSorterTest.java]: replaced containsExactly assertion with less strict startsWith and endsWith assertions.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 10 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: a fake class for testing purpose ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 12 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: removed useless map.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 2 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: Ok, I understand. Please add a comment about the method name then, I think that it's not obvious what a build chain is

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java (hunk 6 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: ok

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java (hunk 8 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxPathSorter.java]: What don't you get ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 10 above)
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: then it contains to many things, I'm sure you only need two fields : one for the value, one for the parent relationship.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxHierarchySorter.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.james.jmap.model.mailbox.Mailbox;
+
+import com.google.common.collect.Lists;
+
+public class MailboxHierarchySorter {
+
+    public List<Mailbox> sortFromRootToLeaf(List<Mailbox> mailboxes) {
+
+        Map<String, Mailbox> mapOfMailboxesById = mailboxes.stream()
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxHierarchySorter.java]: you could extract this part into a indexMailboxesById method I guess

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java [hunk 17]
```diff
@@ -1,54 +1,53 @@
...
+        mailboxes.stream().forEach(graph::registerItem);
 
-        DependencyGraph<FakeCommitNode> graph = new DependencyGraph<>(FakeCommitNode::getHash, m -> m.getParent().orElse(null));
-        mapOfMailboxesById.values().stream().forEach(graph::registerItem);
-
-        List<String> orderedMailboxes = graph.getOrdered()
-                .map(FakeCommitNode::getMessage)
-                .map(o -> o.orElse("(null)"))
+        List<String> orderedMailboxes = graph.getBuildChain()
+                .map(Commit::getMessage)
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: extraction should be separated from action

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java [hunk 18]
```diff
@@ -56,52 +55,53 @@ public void getOrderedShouldReturnOrderedMailbox() {
...
+        Commit(String message) {
+            this(message, null);
         }
 
         @VisibleForTesting
-        FakeCommitNode(Optional<String> message, Optional<FakeCommitNode> parent) {
-            this.hash = DigestUtils.sha1Hex(message.orElse("(empty)"));
+        Commit(String message, Commit parent) {
+            Preconditions.checkArgument(message != null);
+            this.hash = DigestUtils.sha1Hex(message);
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: why do we need a hash _at all_ ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxHierarchySorter.java (hunk 16 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxHierarchySorter.java]: extracted

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java [hunk 19]
```diff
@@ -26,36 +28,36 @@
...
 
-    public Stream<T> getOrdered() {
+    public Stream<T> getBuildChain() {
         DefaultDirectedGraph<T, DefaultEdge> graph = builder.build();
-        AbstractGraphIterator<T, DefaultEdge> i = new TopologicalOrderIterator<>(graph);
-        Iterable<T> iterable = () -> i;
+        return getStreamFromIterator(new TopologicalOrderIterator<>(graph));
+    }
+
+    private static <T> Stream<T> getStreamFromIterator(Iterator<T> iterator) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: Can you put this into an (existing?) Util class? It's not the first time we need this.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 17 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: fixed

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 18 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: removed

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java (hunk 19 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java]: OK, moved to org.apache.james.util.streams.Iterators with (simple!) tests.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 10 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: removed hash

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java (hunk 13 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/DependencyGraphTest.java]: added both tests

**Comment by rouazana** [comment]
:+1:

**Comment by chibenwa** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

**Comment by aduprat** [comment]
merged

**Pull request closed without merging**